Fix and reintroduce initial implementation of CSS Grid Masonry.

sammygill opened the pull request:
#### aeadced20ec62687c9c3f5f1f40a405f8749ab83
<pre>
Fix and reintroduce initial implementation of CSS Grid Masonry.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248475">https://bugs.webkit.org/show_bug.cgi?id=248475</a>
rdar://102772770

Reviewed by Brent Fulgham.

This patch reintroduces the initial implementation of CSS Grid Masonry
after the original one was incorrectly merged in. This addresses the
build issue that was occurring along with some feedback addressed in
the original PR.

Below is the original description from the original patch.

CSS Grid Module Level 3 defines a new layout mode for grid containers
called Masonry Layout. Masonry layout is triggered by setting either
grid-template-rows or grid-template-columns to “masonry.” The axis that
is set to “masonry,” will be called the masonry axis and the other axis
will be known as the grid axis. The algorithm will go through each item
that needs a position and choose the grid axis track that has the most
available room. The specification does not provide any strict
requirements for the masonry axis track (e.g. how many there should be
or when they should be created), so the initial implementation has only
one masonry axis track.

Masonry layout is performed after an initial layout of the grid. The
initial layout will be done as if the masonry axis had “none,” set as
its value. The initial setup for masonry layout is done in the end of
RenderGrid::placeItemsOnGrid after the initial layout has completed.
This gives us the opportunity to collect items into a HashMap or an
appropriate Vector that will be used in a particular part of the layout
algorithm.

The HashMap will contain grid items that were placed on the first
implicit line in the masonry axis and their GridAreas so that they can
be easily placed in the original positions during the first step of the
algorithm. The number of grid axis tracks that are available for this
initial placement should be equal to the number of them before auto
placement of grid items occurs in the original item placement algorithm.

The two Vectors are used to store the items which have a definite grid
axis track placement and items with an indefinite placement, since those
will need to be handled separately.

Once the grid items have been collected, control is given over to the
GridMasonryLayout class that holds all of the state and information for
masonry layout to actually position the items into the appropriate areas.
GridMasonryLayout will hold information such as the running positions of
each of the tracks to reference during the positioning algorithm. These
running positions will be held inside of a Vector that will be
referenced using the corresponding starting line index of the track.
Each item will also have a corresponding offset that is stored in a
HashMap to be used to position at the correct location in its track.

* LayoutTests/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/Grid.cpp:
(WebCore::Grid::setupGridForMasonryLayout):
* Source/WebCore/rendering/Grid.h:
* Source/WebCore/rendering/GridMasonryLayout.cpp: Added.
(WebCore::GridMasonryLayout::performMasonryPlacement):
(WebCore::GridMasonryLayout::resizeAndResetRunningPositions):
(WebCore::GridMasonryLayout::addItemsToFirstTrack):
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
(WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
(WebCore::GridMasonryLayout::masonryAxisMarginBoxForItem):
(WebCore::GridMasonryLayout::updateRunningPositions):
(WebCore::GridMasonryLayout::updateItemOffset):
(WebCore::GridMasonryLayout::nextMasonryPositionForItem):
(WebCore::GridMasonryLayout::offsetForChild const):
* Source/WebCore/rendering/GridMasonryLayout.h: Added.
(WebCore::GridMasonryLayout::GridMasonryLayout):
(WebCore::GridMasonryLayout::gridContentSize const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeTrackBasedSize const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::RenderGrid):
(WebCore::isMasonryRows):
(WebCore::isMasonryColumns):
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::allocateSpaceForMasonryVectors):
(WebCore::RenderGrid::masonryContentSize const):
(WebCore::itemGridAreaStartsAtFirstLine):
(WebCore::RenderGrid::hasDefiniteGridAxisPosition const):
(WebCore::itemGridAreaIsWithinImplicitGrid):
(WebCore::RenderGrid::collectMasonryItems const):
(WebCore::RenderGrid::layoutGridItems):
(WebCore::RenderGrid::gridAreaBreadthForChildIncludingAlignmentOffsets const):
(WebCore::RenderGrid::columnAxisOffsetForChild const):
(WebCore::RenderGrid::rowAxisOffsetForChild const):
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/style/GridArea.h:
(WebCore::GridSpan::masonryAxisTranslatedDefiniteGridSpan):
(WebCore::GridSpan::GridSpan):
* Source/WebCore/rendering/style/GridPositionsResolver.h:
(WebCore::GridPositionsResolver::gridAxisDirection):
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::setColumns):

Canonical link: <a href="https://commits.webkit.org/257185@main">https://commits.webkit.org/257185@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f6896c7ad26fe5a68f6bbb22c0505a3806bcbb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98074 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107536 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167804 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7778 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36053 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90688 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104148 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103726 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84673 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32989 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89431 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1264 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1229 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22368 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4949 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6089 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2529 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41773 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/rendering/Grid.cpp:
```diff
@@ -186,6 +186,13 @@ GridSpan Grid::gridItemSpan(const RenderBox& gridItem, GridTrackSizingDirection
 
+void Grid::setupGridForMasonryLayout()
+{
+    // FIXME: See if we can resize grid instead of clearing it here: https://bugs.webkit.org/show_bug.cgi?id=248472
```
Nit:

`// FIXME(248472): See if we can resize ... etc.`

brentfulgham commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -0,0 +1,167 @@
+    if (m_gridAxisDirection == ForColumns)
+        child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns));
+    else
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
```
Still curious why this isn't just:

`
child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, m_gridAxisDirection));
`

sammygill replied:
That works if the grid axis is in the column direction because we want to update the width. If it's in the row direction we need to update the height instead. (Or are you talking more about replacing the `ForColumns`/`ForRows` with `m_gridAxisDirection`)

brentfulgham commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -0,0 +1,167 @@
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, const GridArea& area)
```
If we never nullptr-check child, could it be passed as a RenderBox&?

brentfulgham commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -0,0 +1,167 @@
+            marginBoxSize = child->logicalWidth() + child->marginLogicalWidth();
+    }
+    return marginBoxSize;
+}
```
Nit: Missing a newline here.

brentfulgham commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -0,0 +1,167 @@
+LayoutUnit GridMasonryLayout::offsetForChild(const RenderBox& child) const
+{
+    if (m_itemOffsets.contains(&child))
+        return m_itemOffsets.get(&child);
```
Nit: This pattern causes two lookups in the hash table (one for 'contains', another for 'get'). If you want to use the result of the lookup, it's better to use 'find':

```
const auto& offsetIter = m_itemOffsets.find(&child);
if (offsetIter == m_itemOffsets.end())
    return 0_lu;
return *offsetIter;
```

There might be a` findOr<>` thing that would work, too.

brentfulgham commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1070,7 +1156,6 @@ void RenderGrid::layoutGridItems()
                 prepareChildForPositionedLayout(*child);
             continue;
         }
-
```
You shouldn't remove this whitespace.

brentfulgham reviewed (approved):
I found a few more small things. no need for another review to land. Or, you can handle them in a follow-up patch if you like.

sammygill committed:
Fix and reintroduce initial implementation of CSS Grid Masonry.
https://bugs.webkit.org/show_bug.cgi?id=248475
rdar://102772770

Reviewed by Brent Fulgham.

This patch reintroduces the initial implementation of CSS Grid Masonry
after the original one was incorrectly merged in. This addresses the
build issue that was occurring along with some feedback addressed in
the original PR.

Below is the original description from the original patch.

CSS Grid Module Level 3 defines a new layout mode for grid containers
called Masonry Layout. Masonry layout is triggered by setting either
grid-template-rows or grid-template-columns to “masonry.” The axis that
is set to “masonry,” will be called the masonry axis and the other axis
will be known as the grid axis. The algorithm will go through each item
that needs a position and choose the grid axis track that has the most
available room. The specification does not provide any strict
requirements for the masonry axis track (e.g. how many there should be
or when they should be created), so the initial implementation has only
one masonry axis track.

Masonry layout is performed after an initial layout of the grid. The
initial layout will be done as if the masonry axis had “none,” set as
its value. The initial setup for masonry layout is done in the end of
RenderGrid::placeItemsOnGrid after the initial layout has completed.
This gives us the opportunity to collect items into a HashMap or an
appropriate Vector that will be used in a particular part of the layout
algorithm.

The HashMap will contain grid items that were placed on the first
implicit line in the masonry axis and their GridAreas so that they can
be easily placed in the original positions during the first step of the
algorithm. The number of grid axis tracks that are available for this
initial placement should be equal to the number of them before auto
placement of grid items occurs in the original item placement algorithm.

The two Vectors are used to store the items which have a definite grid
axis track placement and items with an indefinite placement, since those
will need to be handled separately.

Once the grid items have been collected, control is given over to the
GridMasonryLayout class that holds all of the state and information for
masonry layout to actually position the items into the appropriate areas.
GridMasonryLayout will hold information such as the running positions of
each of the tracks to reference during the positioning algorithm. These
running positions will be held inside of a Vector that will be
referenced using the corresponding starting line index of the track.
Each item will also have a corresponding offset that is stored in a
HashMap to be used to position at the correct location in its track.

* LayoutTests/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/Grid.cpp:
(WebCore::Grid::setupGridForMasonryLayout):
* Source/WebCore/rendering/Grid.h:
* Source/WebCore/rendering/GridMasonryLayout.cpp: Added.
(WebCore::GridMasonryLayout::performMasonryPlacement):
(WebCore::GridMasonryLayout::resizeAndResetRunningPositions):
(WebCore::GridMasonryLayout::addItemsToFirstTrack):
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
(WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
(WebCore::GridMasonryLayout::masonryAxisMarginBoxForItem):
(WebCore::GridMasonryLayout::updateRunningPositions):
(WebCore::GridMasonryLayout::updateItemOffset):
(WebCore::GridMasonryLayout::nextMasonryPositionForItem):
(WebCore::GridMasonryLayout::offsetForChild const):
* Source/WebCore/rendering/GridMasonryLayout.h: Added.
(WebCore::GridMasonryLayout::GridMasonryLayout):
(WebCore::GridMasonryLayout::gridContentSize const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeTrackBasedSize const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::RenderGrid):
(WebCore::isMasonryRows):
(WebCore::isMasonryColumns):
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::allocateSpaceForMasonryVectors):
(WebCore::RenderGrid::masonryContentSize const):
(WebCore::itemGridAreaStartsAtFirstLine):
(WebCore::RenderGrid::hasDefiniteGridAxisPosition const):
(WebCore::itemGridAreaIsWithinImplicitGrid):
(WebCore::RenderGrid::collectMasonryItems const):
(WebCore::RenderGrid::layoutGridItems):
(WebCore::RenderGrid::gridAreaBreadthForChildIncludingAlignmentOffsets const):
(WebCore::RenderGrid::columnAxisOffsetForChild const):
(WebCore::RenderGrid::rowAxisOffsetForChild const):
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/style/GridArea.h:
(WebCore::GridSpan::masonryAxisTranslatedDefiniteGridSpan):
(WebCore::GridSpan::GridSpan):
* Source/WebCore/rendering/style/GridPositionsResolver.h:
(WebCore::GridPositionsResolver::gridAxisDirection):
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::setColumns):

Canonical link: https://commits.webkit.org/257185@main
